### ISSUE TITLE
Failed to compile the emulator - g++ version for `v1.6` release ?

### ISSUE THREAD

**Issue by pcotret** [chipsalliance/rocket-chip#3225]
**Type of issue**: bug report 

**If the current behavior is a bug, please provide the steps to reproduce the problem:**
```bash
git clone https://github.com/chipsalliance/rocket-chip
cd rocket-chip
git checkout v1.6
git submodule update --init
#(compile rocket-tools)
cd emulator
make
```

**What is the current behavior?**
 I'm working with the `v1.6` release. I was trying to compile the emulator (setup given below).
 - I had to change a flag in the emulator Makefile: https://github.com/chipsalliance/rocket-chip/pull/3194 (not merged yet when `v1.6` was released).

Error log while compiling the emulator:
```
g++  -I.  -MMD -I/home/test/Documents/rocket-chip/emulator/verilator/install/share/verilator/include -I/home/test/Documents/rocket-chip/emulator/verilator/install/share/verilator/include/vltstd -DVM_COVERAGE=0 -DVM_SC=0 -DVM_TRACE=0 -faligned-new -fcf-protection=none -Wno-bool-operation -Wno-sign-compare -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-parameter -Wno-unused-variable -Wno-shadow     -O1 -std=c++17 -I/opt/riscv_rocket/include -DVERILATOR -DTEST_HARNESS=VTestHarness -include /home/test/Documents/rocket-chip/src/main/resources/csrc/verilator.h -include /home/test/Documents/rocket-chip/emulator/generated-src/freechips.rocketchip.system.DefaultConfig.plusArgs -I/home/test/Documents/rocket-chip/emulator/generated-src -include /home/test/Documents/rocket-chip/emulator/generated-src/freechips.rocketchip.system.DefaultConfig/VTestHarness.h  -DVL_THREADED -std=gnu++14  -c -o SimDTM.o /home/test/Documents/rocket-chip/src/main/resources/csrc/SimDTM.cc
In file included from /opt/riscv_rocket/include/fesvr/memif.h:10,
                 from /opt/riscv_rocket/include/fesvr/htif.h:6,
                 from /opt/riscv_rocket/include/fesvr/dtm.h:4,
                 from /home/test/Documents/rocket-chip/src/main/resources/csrc/SimDTM.cc:3:
/opt/riscv_rocket/include/fesvr/../riscv/cfg.h:81:8: error: ‘optional’ in namespace ‘std’ does not name a template type
   81 |   std::optional<reg_t>               start_pc;
      |        ^~~~~~~~
/opt/riscv_rocket/include/fesvr/../riscv/cfg.h:81:3: note: ‘std::optional’ is only available from C++17 onwards
   81 |   std::optional<reg_t>               start_pc;
      |   ^~~
make[1]: *** [VTestHarness.mk:65: SimDTM.o] Error 1
```

Questions :
- Is there a way to know which g++ version is recommended?

- Maybe we could also update the [verilator hash](https://github.com/chipsalliance/rocket-chip/blob/master/verilator.hash)? Verilator v4.028 seems to require a gnu++14 compatible compiler.

**What is the expected behavior?**
Being able to compile the emulator.

**Please tell us about your environment:**
- Ubuntu 20.04.5 LTS
- g++ :
```bash
g++ -v       
Using built-in specs.
COLLECT_GCC=g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/9/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:hsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Ubuntu 9.4.0-1ubuntu1~20.04.1' --with-bugurl=file:///usr/share/doc/gcc-9/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,gm2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-9 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-9-Av3uEd/gcc-9-9.4.0/debian/tmp-nvptx/usr,hsa --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1) 
```

**Comment by michael-etzkorn** [comment]
The g++ version you linked doesn't seem to be problematic (that's what I use). Which version of riscv-tools are you on? 
v1.6 points to this hash: e2c6d1577a75f506fe992c3eb20a75174504476e. 

My fesvr doesn't include a cfg.h file? If I had a guess the version you're on is beyond what v1.6 supports. I use a toolchain built from chipyard v1.8 (which is compatible with RC v1.6).

**Comment by michael-etzkorn** [comment]
That being said, this issue brings up a good point about documenting in the changelog the g++ and riscv-tools support.

**Comment by pcotret** [comment]
...
It must be that, I was using the `master` branch without using the hash given in https://github.com/chipsalliance/rocket-chip/blob/1.6/riscv-tools.hash.

I'll try it later.

**Comment by pcotret** [comment]
I had to update several submodules URL (my `git` didn't accept `git://<repo_url>` anymore) but it finally worked.

Here is my script to compile rocket-tools for the `v1.6` release: https://gist.github.com/pcotret/11afe52a1834172981c1e371b8bdcf03

**Comment by michael-etzkorn** [comment]
Should be resolved by #3226, but we'll need to introduce a backport flow to fix v1.6 on checkout. I'll make a note about rocket-tools compatibility in the next changelog.
